src/backend/distributed/ddl/foreign_constraint.c

```diff
@@ -21,11 +22,588 @@
+ * by relation C, then the result list for relation A consists of the relation
+ * IDs of relation B and relation C.
+ */
+List *
+ReferencedRelationIdList(Oid relationId)
+{
+	return GetReferencedRelationIdHelper(relationId, false);
+}
+
+
+/*
+ * ReferencingRelationIdList is a wrapper function around GetRefenceRelationIdHelper
+ * to get list of relation IDs which are referencing by the given relation id.
+ * Note that, if relation A is referenced by relation B and relation B is referenced
+ * by relation C, then the result list for relation C consists of the relation
+ * IDs of relation A and relation B.
+ */
+List *
+ReferencingRelationIdList(Oid relationId)
+{
+	return GetReferencedRelationIdHelper(relationId, true);
+}
+
+
+/*
+ * GetReferencedRelationIdHelper returns the list of oids affected or affecting
+ * given relation id. It uses the transitivity matrix which is computed once in
+ * a session. It is a helper function for providing results to public functions
+ * ReferencedRelationIdList and ReferencingRelationIdList.
+ */
+static List *
+GetReferencedRelationIdHelper(Oid relationId, bool isAffecting)
+{
+	List *foreignKeyList = NIL;
+	bool isFound = false;
+	FRelNode *relationNode = NULL;
+	uint32 relationIndex = -1;
+	MemoryContext oldContext = MemoryContextSwitchTo(CacheMemoryContext);
+
+	CreateForeignKeyRelationGraph();
+	relationNode = (FRelNode *) hash_search(frelGraph->nodeMap, &relationId,
+											HASH_FIND, &isFound);
+	if (!isFound)
+	{
+		/*
+		 * If there is no node with the given relation id, that means given table
+		 * does not referencing and does not referenced by any table
+		 */
+		MemoryContextSwitchTo(oldContext);
+		return NIL;
+	}
+	else
+	{
+		/*
+		 * Create the returning list according to the direction of reference. We
+		 * utilize indexes of each relation node (which has relation id as an
+		 * attribute) to get result from transitivity matrix easily.
+		 */
+		relationIndex = relationNode->index;
+		if (isAffecting)
+		{
+			uint32 tableCounter = 0;
+
+			while (tableCounter < frelGraph->nodeCount)
+			{
+				if (frelGraph->transitivityMatrix[relationIndex][tableCounter])
+				{
+					Oid referredTableOid = frelGraph->indexToOidArray[tableCounter];
+					foreignKeyList = lappend_oid(foreignKeyList, referredTableOid);
+				}
+
+				tableCounter += 1;
+			}
+		}
+		else
+		{
+			uint32 tableCounter = 0;
+
+			while (tableCounter < frelGraph->nodeCount)
+			{
+				if (frelGraph->transitivityMatrix[tableCounter][relationIndex])
+				{
+					Oid referringTableOid = frelGraph->indexToOidArray[tableCounter];
+					foreignKeyList = lappend_oid(foreignKeyList, referringTableOid);
+				}
+
+				tableCounter += 1;
+			}
+		}
+	}
+
+	MemoryContextSwitchTo(oldContext);
+
+	return foreignKeyList;
+}
+
+
+/*
+ * CreateForeignKeyRelationGraph creates the foreign key relation graph using
+ * foreign constraint provided by pg_constraint metadata table.
+ */
+static void
```

Comment:
_Everything_ inside this function will be allocated in `CacheMemoryContext`, including anything called during traversal of the `pg_constraint` catalog.

Rather than switching into `CacheMemoryContext`, it might be best to create some sort of `CacheMemoryContext` child specific to this code once and then use it everywhere you need long-lived allocations.